doc: debugging: Introduce guide covering debug probes and host tools

MaureenHelm opened the pull request:
Zephyr boards use a wide range of debug probes and debug host tools that
can often be confusing to users. Introduce a new debugging guide that
documents all the different variations of debug probes and host tools in
one place, including which combinations are valid.

@carlescufi, @erwango what do you think about including sections for ST-LINK and and Nordic here too?

Fixes #13584

MaureenHelm committed:
doc: debugging: Introduce guide covering debug probes and host tools

Zephyr boards use a wide range of debug probes and debug host tools that
can often be confusing to users. Introduce a new debugging guide that
documents all the different variations of debug probes and host tools in
one place, including which combinations are valid.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: mimxrt10{20,50}_evk: Rework programming and debugging doc

Reworks the programming and debugging section in the mimxrt1020_evk and
mimxrt1050_evk board documents to leverage the new debugging guide
covering debug probes and host tools.

These boards support both the OpenSDA J-Link onboard debug probe and the
J-Link external debug probe.

Corrects an error linking to the wrong OpenSDA J-Link firmware.

Updates the flashing section to reflect that the ``flash`` build system
target is now supported.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

dbkinder commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+LPC-Link2 DAPLink Onboard Debug Probe
+*************************************
+
+TODO
```
Should we wait for these TODO sections before merging?

dbkinder reviewed (dismissed):
Looks just fine so far... a couple of TODO sections left.

ulfalizer commented on doc/guides/debugging/host-tools.rst:
```diff
@@ -0,0 +1,68 @@
+J-Link GDB Server and Commander, and to install the associated USB device
+drivers. These two tools are integrated into the Zephyr build system to provide
+``west debug`` and ``west flash`` support respectively. RTT Viewer and
+SystemView can be downloaded separately, but not are not required.
```
"not are not"

mbolivar commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+implementations: NXP LPC boards have LPC-Link2, NXP Kinetis (former Freescale)
+boards have OpenSDA, and ST boards have ST-LINK. Each onboard debug probe
+microcontroller can support one or more types of firmware that communicate with
+their respective debug host tools. For example, an OpenSDA microcontroller can
```
s/their/its/

MaureenHelm replied:
I don't think so. "One or more types of firmware" -> their.

mbolivar replied:
Oh, I read the antecedent as "microcontroller", not "types". Makes sense now.

mbolivar commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+boards have OpenSDA, and ST boards have ST-LINK. Each onboard debug probe
+microcontroller can support one or more types of firmware that communicate with
+their respective debug host tools. For example, an OpenSDA microcontroller can
+be programmed with DAPLink firmware to communicate with pyOCD or openOCD debug
```
s/openOCD/OpenOCD/

mbolivar commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+host tools, or with J-Link firmware to communicate with J-Link debug host
+tools.
+
+Some supported boards in Zephyr do not include an onboard debug probe and
```
How about:

```
Boards which do not have integrated debug hardware require an additional external debug
probe to use these features. Usually, these boards have a JTAG or SWD header which can be
connected to the external debug probe via a cable or individual wires.

You can also sometimes use an external debug probe with boards that have integrated debug
hardware. One reason this may be useful is that the integrated hardware may have limitations,
such as lack of support for advanced debuggers or high-speed tracing. You may need to adjust
jumpers on the board to allow debugging via the external probe without interference from the
integrated hardware.
```

mbolivar commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+.. _debug-probes:
+
+Debug Probes
+############
```
Introductory text defining a debug probe and its connections in the system would be nice to have. Something like:

```
A *debug probe* is special hardware which allows you to control execution
of a Zephyr application running on a separate board. Debug probes usually
allow reading and writing registers and memory, and support breakpoint
debugging of the Zephyr application on your host workstation using tools like
GDB. They may also support other debug software and more advanced features
such as :ref:`tracing program execution <tracing>`. For details on the related
host software supported by Zephyr, see :ref:`debug-host-tools`.

Debug probes are usually connected to your host workstation via USB; they
are sometimes also accessible via an IP network or other means. They usually
connect to the device running Zephyr using the JTAG or SWD protocols. Debug
probes are either separate hardware devices or circuitry integrated into the same
board which runs Zephyr.
```

mbolivar reviewed (commented):
This is great! Thanks @MaureenHelm. I've wanted to write a doc like this for a while but never found the time. I have some suggestions for next steps.

MaureenHelm commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,143 @@
+Debug Probes
+############
+
+Many supported boards in Zephyr include a second microcontroller that serves as
```
Maybe it's because I work for an IC company, but when I see "integrated" like this I think integrated into the SoC. I want to emphasize the fact that there is a second MCU and firmware associated it (and maybe someday it will run zephyr too!)

mbolivar replied:
Heh, fair. How about "on-board"?

erwango commented:
@MaureenHelm , is this replacement of doc/guides/tools ?

MaureenHelm commented:
> @MaureenHelm , is this replacement of doc/guides/tools ?

Yes

MaureenHelm committed:
boards: mimxrt10{60,64}_evk: Rework programming and debugging doc

Reworks the programming and debugging section in the mimxrt1060_evk and
mimxrt1064_evk board documents to leverage the new debugging guide
covering debug probes and host tools.

Neither of these boards have OpenSDA J-Link board-specific firmware,
therefore the only debug probe currently supported is the external
J-Link.

Updates the flashing section to reflect that the ``flash`` build system
target is now supported.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: frdm_{kl25z, k64f, kw41z}: Rework programming and debugging doc

Reworks the programming and debugging section in the freedom board
documents to leverage the new debugging guide covering debug probes and
host tools.

These boards support OpenSDA DAPLink and OpenSDA J-Link onboard debug
probes. They can also support an external J-Link probe, but this
requires board modifications (cutting traces) and is therefore not
documented.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: frdm_kw41z: Default to daplink firmware and pyocd runner

The frdm_kw41z board was originally configured in zephyr to use the
jlink runner by default because pyocd didn't yet support the kw41z soc.
Support for kw41z was added in pyocd v0.9.0, so we can now default to
daplink firmware and pyocd.

Now all freedom boards in zephyr consistently use daplink and pyocd.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: hexiwear_k64f: Default to jlink firmware and jlink runner

The hexiwear_k64f and hexiwear_kw40z share the same OpenSDA
microcontroller on the hexiwear docking station, so make the jlink the
default firmware on both for consistency. We use jlink instead of
daplink by default because hexiwear_kw40z can only use Segger RTT for
the console (there is no UART available).

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: hexiwear_{k64, kw40z}: Rework programming and debugging doc

Reworks the programming and debugging section in the hexiwear board
documents to leverage the new debugging guide covering debug probes and
host tools.

Recommends the OpenSDA J-Link debug probe because the k64 and kw40z
share the same OpenSDA microcontroller, and the kw40z requires Segger
RTT for a console. It is possible to use daplink firmware for the k64,
however it is not recommended because it requires switching the firmware
back to jlink to access the kw40z.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: lpcxpresso54114: Rework programming and debugging doc

Reworks the programming and debugging section in the lpcxpresso54114
board document to leverage the new debugging guide covering debug probes
and host tools.

This board supports the LPC-Link2 J-Link onboard debug probe.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: usb_kw24d512: Rework programming and debugging doc

Reworks the programming and debugging section in the usb_kw24d512 board
document to leverage the new debugging guide covering debug probes and
host tools.

This board does not have an OpenSDA microcontroller, therefore the only
debug probe currently supported is the external J-Link.

Updates the flashing section to reflect that the ``flash`` build system
target is now supported.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
doc: tools: Remove opensda page

This page has been replaced by doc/guides/debugging/probes.rst and
doc/guides/debugging/host-tools.rst

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: 96b_nitrogen: Update link to pyocd host tools

Updates the 96b_nitrogen board document to link to the new debugging
guide instead of the opensda page.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
@dbkinder @mbolivar Please revisit. I filled in all the previously marked TODO sections, and moved the PR out of draft state.

erwango reviewed (approved):
LGTM, ST-Link add-ons pushed in #14747

MaureenHelm commented on boards/arm/frdm_k64f/doc/index.rst:
```diff
@@ -202,64 +202,100 @@ Only USB device function is supported in Zephyr at the moment.
+
+Follow the instructions in :ref:`opensda-jlink-onboard-debug-probe` to program
+the `OpenSDA J-Link Generic Firmware for V3.2 Bootloader`_. Note that Segger
+does provide an OpenSDA J-Link Board-Specific Firmware for this board, however
```
Yes, but it doesn't work with the updated bootloader.

mbolivar reviewed (approved):
All my comments are addressed; sorry for the delay!

dbkinder commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,170 @@
+This probe is realized by programming the OpenSDA microcontroller with DAPLink
+OpenSDA firmware. NXP provides `OpenSDA DAPLink Board-Specific Firmwares`_.
+
+Before you program the firmware, make sure to install the debug host tools
```
This would read better as:
```
Install the debug host tools before you program the firmware.
```

dbkinder commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,170 @@
+boards to support their external flash memories, whereas generic firmwares are
+compatible with all Kinetis boards.
+
+Before you program the firmware, make sure to install the debug host tools
```
Same here. This would read better as:
```
Install the debug host tools before you program the firmware.
```

dbkinder commented on doc/guides/debugging/probes.rst:
```diff
@@ -0,0 +1,170 @@
+- :ref:`jlink-debug-host-tools`
+- :ref:`openocd-debug-host-tools`
+
+Before you use this debug probe, make sure to install the debug host tools
```
One more time. This would read better as:
```
Install the debug host tools before you use this debug probe.
```

dbkinder reviewed (changes requested):
Looks good, though there seems to be material repeated in different docs.  It's not always the case that the words are the same though, so trying to create an include file that we can plug into each spot might not work.  Otherwise, just a few working tweaks...

dbkinder commented:
Hmm @nashif, you requested my review and then merged the PR... oh well.

nashif commented:
> Hmm @nashif, you requested my review and then merged the PR... oh well.

oops, sorry.
